File: web/lib/rooms.ts
// Rooms domain client: REST for room/member/message CRUD plus the realtime
// WebSocket at /api/v1/rooms/ws. Messages are sent over the socket (the server
// persists them and fans the event out to every member, sender included), so
// the room page only ever appends events it receives.

import { api, API_BASE_URL } from "./api";
import type { Room, RoomMember, RoomMessage } from "./types";

// ------------------------------------------------------------------ rooms

export function listRooms(signal?: AbortSignal): Promise<Room[]> {
  return api.get<Room[]>("/api/v1/rooms", signal);
}

export function getRoom(id: string): Promise<Room> {
  return api.get<Room>(`/api/v1/rooms/${id}`);
}

export function createRoom(body: {
  name: string;
  description?: string;
  is_public?: boolean;
}): Promise<Room> {
  return api.post<Room>("/api/v1/rooms", body);
}

export function updateRoom(
  id: string,
  body: { name?: string; description?: string; is_public?: boolean },
): Promise<Room> {
  return api.patch<Room>(`/api/v1/rooms/${id}`, body);
}

export function deleteRoom(id: string): Promise<{ ok: boolean }> {
  return api.delete<{ ok: boolean }>(`/api/v1/rooms/${id}`);
}

// --------------------------------------------------------------- membership

export function joinRoom(id: string): Promise<Room> {
  return api.post<Room>(`/api/v1/rooms/${id}/join`);
}

export function leaveRoom(id: string): Promise<{ ok: boolean }> {
  return api.post<{ ok: boolean }>(`/api/v1/rooms/${id}/leave`);
}

export function listMembers(id: string): Promise<RoomMember[]> {
  return api.get<RoomMember[]>(`/api/v1/rooms/${id}/members`);
}

export function addMember(
  id: string,
  body: { user_id?: string; email?: string },
): Promise<RoomMember> {
  return api.post<RoomMember>(`/api/v1/rooms/${id}/members`, body);
}

export function removeMember(id: string, userId: string): Promise<{ ok: boolean }> {
  return api.delete<{ ok: boolean }>(`/api/v1/rooms/${id}/members/${userId}`);
}

export function setMemberRole(
  id: string,
  userId: string,
  role: "owner" | "member",
): Promise<RoomMember> {
  return api.patch<RoomMember>(`/api/v1/rooms/${id}/members/${userId}`, { role });
}

// ---------------------------------------------------------------- messages

export function listMessages(
  id: string,
  opts: { beforeId?: string; limit?: number } = {},
): Promise<RoomMessage[]> {
  const q = new URLSearchParams();
  if (opts.beforeId) q.set("before_id", opts.beforeId);
  if (opts.limit) q.set("limit", String(opts.limit));
  const suffix = q.toString() ? `?${q}` : "";
  return api.get<RoomMessage[]>(`/api/v1/rooms/${id}/messages${suffix}`);
}

export function createMessageREST(id: string, content: string): Promise<RoomMessage> {
  return api.post<RoomMessage>(`/api/v1/rooms/${id}/messages`, { content });
}

export function deleteMessageREST(id: string, messageId: string): Promise<{ ok: boolean }> {
  return api.delete<{ ok: boolean }>(`/api/v1/rooms/${id}/messages/${messageId}`);
}

// ------------------------------------------------------------------- socket

/**
 * The rooms WebSocket URL. In dev the API lives on its own port, so we swap
 * the scheme of API_BASE_URL; in production the API is same-origin.
 */
export function roomSocketUrl(): string {
  if (API_BASE_URL) {
    return `${API_BASE_URL.replace(/^http/, "ws")}/api/v1/rooms/ws`;
  }
  const proto = typeof location !== "undefined" && location.protocol === "https:" ? "wss" : "ws";
  return `${proto}://${location.host}/api/v1/rooms/ws`;
}

/** Wire events the socket can deliver. */
export type RoomEvent =
  | { type: "members"; room_id: string; users: Array<{ user_id: string; name: string }> }
  | {
      type: "presence";
      room_id: string;
      action: "joined" | "left";
      user: { id: string; name: string };
    }
  | { type: "message"; room_id: string; message: RoomMessage }
  | { type: "typing"; room_id: string; user: { id: string; name: string }; typing: boolean }
  | { type: "error"; message: string };

File: api/jyoti_api/api/rooms.py
"""Rooms endpoints: /api/v1/rooms (REST) + the /ws realtime socket."""

from __future__ import annotations

from typing import Any

from fastapi import APIRouter, WebSocket
from pydantic import BaseModel, Field

from jyoti_api.api.deps import CurrentUser, SessionDep
from jyoti_api.domain import rooms as domain
from jyoti_api.realtime.rooms import rooms_websocket

router = APIRouter(prefix="/api/v1/rooms", tags=["rooms"])


class RoomCreate(BaseModel):
    name: str = Field(min_length=1, max_length=200)
    description: str = Field(default="", max_length=2000)
    is_public: bool = False


class RoomUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=200)
    description: str | None = Field(default=None, max_length=2000)
    is_public: bool | None = None


class MessageCreate(BaseModel):
    content: str = Field(min_length=1, max_length=8000)


class MemberAdd(BaseModel):
    """Add a member by user id, or by email (resolved server-side)."""

    user_id: str | None = None
    email: str | None = Field(default=None, max_length=320)


class RoleUpdate(BaseModel):
    role: str


@router.websocket("/ws")
async def ws(websocket: WebSocket, token: str | None = None) -> None:
    """Live room events. Auth via ``?token=``, the session cookie, or ``?key=``."""
    await rooms_websocket(websocket, token=token)


@router.get("")
def list_rooms(user: CurrentUser, session: SessionDep) -> list[dict[str, Any]]:
    return domain.list_rooms(session, user.id)


@router.post("", status_code=201)
def create_room(body: RoomCreate, user: CurrentUser, session: SessionDep) -> dict[str, Any]:
    room = domain.create_room(session, user.id, body.name, body.description, body.is_public)
    return domain.to_room_dto(session, room, user.id)


@router.get("/{room_id}")
def get_room(room_id: str, user: CurrentUser, session: SessionDep) -> dict[str, Any]:
    return domain.get_room(session, user.id, room_id)


@router.patch("/{room_id}")
def update_room(
    room_id: str, body: RoomUpdate, user: CurrentUser, session: SessionDep
) -> dict[str, Any]:
    room = domain.update_room(
        session, user.id, room_id, body.name, body.description, body.is_public
    )
    return domain.to_room_dto(session, room, user.id)


@router.delete("/{room_id}")
def delete_room(room_id: str, user: CurrentUser, session: SessionDep) -> dict[str, bool]:
    domain.delete_room(session, user.id, room_id)
    return {"ok": True}


# ------------------------------------------------------------------ membership


@router.post("/{room_id}/join")
def join_room(room_id: str, user: CurrentUser, session: SessionDep) -> dict[str, Any]:
    return domain.join_room(session, user.id, room_id)


@router.post("/{room_id}/leave")
def leave_room(room_id: str, user: CurrentUser, session: SessionDep) -> dict[str, bool]:
    domain.leave_room(session, user.id, room_id)
    return {"ok": True}


@router.get("/{room_id}/members")
def list_members(room_id: str, user: CurrentUser, session: SessionDep) -> list[dict[str, Any]]:
    return domain.list_members(session, user.id, room_id)


@router.post("/{room_id}/members", status_code=201)
def add_member(
    room_id: str, body: MemberAdd, user: CurrentUser, session: SessionDep
) -> dict[str, Any]:
    return domain.add_member(session, user.id, room_id, body.user_id, email=body.email)


@router.delete("/{room_id}/members/{target_id}")
def remove_member(
    room_id: str, target_id: str, user: CurrentUser, session: SessionDep
) -> dict[str, bool]:
    domain.remove_member(session, user.id, room_id, target_id)
    return {"ok": True}


@router.patch("/{room_id}/members/{target_id}")
def set_role(
    room_id: str, target_id: str, body: RoleUpdate, user: CurrentUser, session: SessionDep
) -> dict[str, Any]:
    return domain.set_member_role(session, user.id, room_id, target_id, body.role)


# ------------------------------------------------------------------ messages


@router.get("/{room_id}/messages")
def list_messages(
    room_id: str,
    user: CurrentUser,
    session: SessionDep,
    before_id: str | None = None,
    limit: int = 50,
) -> list[dict[str, Any]]:
    return domain.list_messages(session, user.id, room_id, before_id=before_id, limit=limit)


@router.post("/{room_id}/messages", status_code=201)
def create_message(
    room_id: str, body: MessageCreate, user: CurrentUser, session: SessionDep
) -> dict[str, Any]:
    message = domain.create_message(session, user.id, room_id, body.content)
    return domain.to_message_dto(session, message)


@router.delete("/{room_id}/messages/{message_id}")
def delete_message(
    room_id: str, message_id: str, user: CurrentUser, session: SessionDep
) -> dict[str, bool]:
    domain.delete_message(session, user.id, room_id, message_id)
    return {"ok": True}
